### ISSUE TITLE
New Behavior with Block Cloning?

### ISSUE THREAD

**Issue by theriverlethe** [markusressel/zfs-inplace-rebalancing#32]
What effect will block cloning in OpenZFS 2.2 have on this script? The documentation may need to be updated.

**Comment by feld** [comment]
On FreeBSD:

sysctl vfs.zfs.bclone_enabled=0

Linux... unsure how to disable it.

**Comment by markusressel** [comment]
I don't know what block cloning is or how it works, so I don't know what effect it has. If someone wants to figure this out and update the README, feel free to open a PR :)

**Comment by johnpyp** [comment]
I think the impact on the script will be that blocks will default be reference cloned when using `cp` and most normal fs apis, effectively making it so that the rebalancing this script does will have no effect (the data will stay on the same drive). As far as the script goes, by using `--reflink=never` in the cp command, we should be able to mitigate that.

Disabling block cloning on the pool is not the solution (reflinks are awesome and I encourage people to use them).

**Comment by theriverlethe** [comment]
> I think the impact on the script will be that blocks will default be reference cloned when using `cp` and most normal fs apis, effectively making it so that the rebalancing this script does will have no effect (the data will stay on the same drive). As far as the script goes, by using `--reflink=never` in the cp command, we should be able to mitigate that.
> 
> Disabling block cloning on the pool is not the solution (reflinks are awesome and I encourage people to use them).

This is the way.

**Comment by theriverlethe** [comment]
> I don't know what block cloning is or how it works, so I don't know what effect it has. If someone wants to figure this out and update the README, feel free to open a PR :)

I'm not a ZFS expert, but I believe it replaces a dumb copy operation with almost-free, on-the-fly deduplication. So, "copying" a file on the same pool will only create new reference points to the same blocks rather than duplicating the data. If you delete the "copy" in the old location, the new copy will remain linked to the same blocks. I assume the copies only diverge when one of them is modified.

**Comment by Sawtaytoes** [comment]
This explains why all my pool-to-pool copies have been ridiculously fast after moving to OpenZFS 2.2 via TrueNAS SCALE v23.

So all we need is the script to be modified with `cp --reflink=never`, and we're good?

**Comment by johnpyp** [comment]
Should be, yeah.

**Comment by markusressel** [comment]
Does this (new?) flag `--reflink=never` exist on all platforms (linux, freebsd, osx)? I can check on linux, but I cannot on the other platforms.

**Comment by Brandoskey** [comment]
I've been going mad trying to diagnose why no new data is being written to my newly added vdev. Is someone maintaining a fork of the updated script with the above fixes that I can try?

**Comment by johnpyp** [comment]
There's an open PR for it, you can try that for now

**Comment by Brandoskey** [comment]
I'm 99% sure it's unrelated to the script, but my pool metadata was badly corrupted while running the script with --reflink=never added. Just as a sanity check, have others in this thread successfully run the script on a system with block cloning using the latest version?

I left MD5 checks enabled, if that matters

**Comment by markusressel** [comment]
I have never used the script on my own ZFS pools with the `--reflink=never` change. The CI tests is all that were done by "me". The comments within this issue are the sole reason the change was added, and since there was no response to my request about testing on FreeBSD and OSX I only added it for linux.

What exactly tells you that the metadata is corrupt and how? Should it even be possible for this flag to have such an effect at all? As I understand it setting this flag to `always` should simply "revert" `cp` behavior like it has been before block cloning was a thing... right?

**Comment by Brandoskey** [comment]
I doubt it was the script, but wanted to share my experience in the off chance it was and someone else could be spared. When I added my new vdev TrueNAS didn't create the swap partition that was on the other 3 vdevs which could have lead to the corruption.

**Comment by johnpyp** [comment]
There are some issues being tracked by the zfs team that can cause some kinds of corruption when using block cloning (enough to make them turn it off by default temporarily in 2.2.1) - I assume this is likely the culprit in your case. Not familiar with the details, but the script itself isn't doing anything wrong, it's almost certainly just somehow coinciding with those other issues...

**Comment by Brandoskey** [comment]
I've been able to recover much of my data, however anything I ran the rebalance script on shows a lot of corruption.

Not saying it's anything the script does, but maybe it's just exposing the issues in 2.2.0. I also saw this post this morning https://www.reddit.com/r/zfs/comments/1826lgs/psa_its_not_block_cloning_its_a_data_corruption/

When I ran the script, I initially did so without the --reflink=never flag in the script. I didn't run it very long as I could see nothing was being rebalanced. Then I added the flag to the script manually. Maybe something with me manually adding the flag caused my issue, or I just ran into the perfect circumstances to self destruct my pool.

Anyhow, until they release some bug fixes for this latest version of zfs, I'll not be attempting to rebalance just to be safe. The good news is I'll need to rebuild the pool anyhow so a rebalance won't be needed.

I kept md5 checks enabled and still got massive corruption.

If you're on zfs 2.2.0, maybe just hold off on running any sort of rebalance until they patch these bugs.
